spamc.go

```diff
@@ -67,10 +68,148 @@ func (c *Client) send(
 	return conn, nil
 }
 
+// rearrangeMessageReader will rearrange the message reader to make sure SA will understand the message
+func (c Client) rearrangeMessageReader(message io.ReadSeeker, size int64) (io.Reader, int64, error) {
+	var headers bytes.Buffer
+	var rest bytes.Buffer
+
+	isASCII := func(m []byte) bool {
+		for _, c := range m {
+			if c > 127 {
+				return false
+			}
+		}
+		return true
+	}
+
+	r := bufio.NewReader(message)
+
+	var returnError error
+
+	// we assume that there is no newline between headers and body, so increment size
+	size += 2
+
+	for {
+		chars, err := r.Peek(2)
+		if err != nil {
+			returnError = err
+			break
+		}
+		if bytes.Equal(chars, []byte("\r\n")) {
+			// we were wrong there was a newline between headers and body
+			// decrement size
+			size -= 2
+			// read and discard it, and we are done
+			_, err = r.Read(make([]byte, 2))
+			if err != nil {
+				returnError = err
+			}
+			break
+		}
+		if chars[0] == ' ' || chars[0] == '\t' {
+			// if we have a previous key append value, and continue
+			// if not we are not reading headers
+			if headers.Len() > 0 {
+				line, err := r.ReadSlice('\n')
+				if err != nil {
+					rest.Write(line)
+					returnError = err
+					break
+				}
+				if !isASCII(line) {
+					rest.Write(line)
+					returnError = errors.New("non ascii characters found: [" + string(line) + "]")
+					break
+				}
+				headers.Write(line)
+				continue
+			} else {
+				// this line start with whitespace but there is no
+				// previous header to append, so it's not a header
+				returnError = errors.New("multiline header part found without first header")
+				break
+			}
+		}
+
+		// it should be a header read it
+		line, err := r.ReadSlice('\n')
+
+		if err != nil {
+			rest.Write(line)
+			returnError = err
+			break
+		}
+		// Check for a key, we must have one for it to be a header
+		i := bytes.IndexByte(line, ':')
+		if i < 0 {
+			rest.Write(line)
+			returnError = errors.New("malformed MIME header line: [" + string(line) + "]")
+			break
+		}
+		if !isASCII(line) {
+			rest.Write(line)
+			returnError = errors.New("non ascii characters found: [" + string(line) + "]")
+			break
+		}
+		headers.Write(line)
+	}
+
+	// isTerminated will check last characters of the reader for end of line
+	isTerminated := func(message io.ReadSeeker) (bool, error) {
```

Comment:
Like the `isASCII`, would be better to put this as a regular function.

This is also something we can add a separate test for.

Reply:
Check ;) will do